src/backends/gpu/stack_common.hpp

```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include <backends/fvm_types.hpp>
+
+namespace nest {
+namespace mc {
+namespace gpu {
+
+// stores a single crossing event
+struct threshold_crossing {
+    fvm_size_type index;    // index of variable
+    fvm_value_type time;    // time of crossing
+
+    friend bool operator==(threshold_crossing l, threshold_crossing r) {
+        return l.index==r.index && l.time==r.time;
+    }
+};
+
+// Concrete storage of gpu stack datatype.
+// The stack datatype resides in host memory, and holds a pointer to the
+// stack_storage in managed memory, which can be accessed by both host and
+// gpu code.
+template <typename T>
+struct stack_storage {
+    using value_type = T;
+
+    // The number of items of type value_type that can be stored in the stack
+    unsigned capacity;
+
+    // The number of items that have been stored
+    unsigned size;
```

Comment:
Comment and member name don't quite align with semantics, being the number of attempted stores, or the offset where the store would be performed if within capacity. i.e. `size` is only the size if it is less than or equal to `capacity`.

How about, e.g.
```
    // Offset to next storage slot. Counts number of pushes, and equals size of stack if not greater than capacity.
    unsigned next;
```